.fixtures.yml

```diff
@@ -1,13 +1,9 @@
 fixtures:
   repositories:
     "stdlib":
-      "repo": "git://github.com/puppetlabs/puppetlabs-stdlib.git"
-      "ref": "master"
+      repo: "https://github.com/puppetlabs/puppetlabs-stdlib.git"
```

Comment:
Would it be better to use the forge for everything except hocon?

Reply:
That's an interesting idea; I'm surprised I've never considered it before, and it might be better.   I suppose it depends if your tests are more likely to require on than get broken by unpublished changes in dependency repos' tracking branches.

Not knowing the conventions for voxpup, I downloaded the entire github org and ran a `cat */.fixtures.yml` on all the repos.  As far as I can tell, 100% of the `.fixtures.yml` files us git, so I'll keep that convention during this conversion.

I think you should propose using Forge fixture in Slack, though.  If it's better, it probably should become the general policy.

Reply:
It depends on what you want to achieve. Test against the current version or test against the next version. In theory you should be testing both to get maximum coverage, but I'm not sure it's worth the CI resources.

There are more things to consider. For example, does the version in git satisfy the requirement in `metadata.json`? Would installing from the forge do that.

Reply:
`.fixtures.yml` predates the existence of the forge and `metadata.json`.  If we were starting from scratch today, I think using `metadata.json` would be the default and I hope that is where we end up.  I have thought several times that I should implement that...

Reply:
slightly related; for my private setup I use the follwoing rake task:

```ruby
begin
  desc 'Generate .fixtures.yml from metadata.json + .fixtures-deps.yml'
  task :generate_fixtures do
    #fixtures = MetadataJsonDeps.build_fixtures(File.join(__dir__, 'metadata.json'))
    generated_fixtures = %x(bundle exec generate-fixtures-yaml metadata.json)
    fixtures_deps = File.read(File.join(__dir__, '.fixtures-deps.yml'))
    generated_fixtures_yaml = YAML.safe_load(generated_fixtures)
    fixtures_deps_yaml = YAML.safe_load(fixtures_deps)
    generated_fixtures_yaml['fixtures']['repositories'].merge!(fixtures_deps_yaml['fixtures']['repositories'])
    File.open(File.join(__dir__, '.fixtures.yml'), 'w') {|file| file.puts generated_fixtures_yaml.to_yaml }
  end
rescue LoadError
end
```

I have all direct dependencies in my metadata.json. core modules and dependencies from the dependencies are in fixtures-deps.yml:

```
---
fixtures:
  repositories:
    apt: https://github.com/puppetlabs/puppetlabs-apt
    apache: https://github.com/puppetlabs/puppetlabs-apache
    concat: https://github.com/puppetlabs/puppetlabs-concat
    epel: https://github.com/voxpupuli/puppet-epel
    mcollective: https://github.com/choria-io/puppet-mcollective
    datacat: https://github.com/richardc/puppet-datacat
    elastic_stack: https://github.com/voxpupuli/puppet-elastic_stack
    python: https://github.com/voxpupuli/puppet-python
    ssh_keys: https://github.com/puppetlabs/puppetlabs-sshkeys_core
    yumrepo: https://github.com/puppetlabs/puppetlabs-yumrepo_core
    mount: https://github.com/puppetlabs/puppetlabs-mount_core
    cron: https://github.com/puppetlabs/puppetlabs-cron_core
    augeasproviders_core: https://github.com/voxpupuli/puppet-augeasproviders_core.git
    augeas_core: https://github.com/puppetlabs/puppetlabs-augeas_core.git
    selinux_core: https://github.com/puppetlabs/puppetlabs-selinux_core

```

Reply:
```ruby
    File.open(File.join(__dir__, '.fixtures.yml'), 'w') {|file| file.puts generated_fixtures_yaml.to_yaml }
```

You can simplify this to

```ruby
    File.write(File.join(__dir__, '.fixtures.yml'), generated_fixtures_yaml.to_yaml)
```

Reply:
@bastelfreak I am currently advocating for the poorly named "_core" mods to be listed in `metadata.json` unless our policy is to only support AIO packaging. https://github.com/voxpupuli/puppet-selinux/pull/355  Adopting that policy would probably eliminate the need listing anything in your `.fixutres.yml`.

I think we can [probably] all agree that at this point, `metadata.json` should be the source of truth for puppet module dependencies.

I also believe that we have too many different ways of declaring a inventory of puppet modules.  There is at least `metadata.json`, `.fixtures.yml`, helper methods in `spec_helper_acceptance.rb` and the like and `Puppetfile.  `.fixtures.yml` was created in the era when many were using git submodules to manage env inventory.  There should be room for simplification.  A reasonably path forward may be to try to most closely resemble rubygem machinery using `metadata.json` (`.gemspec`analog) and `Puppetfile` (`Gemfile` analog).

Specifically:
1. All dependencies, include `_core` mods, are declared in `metadata.json` and this list of deps is used by default for both `rspec-puppet` and `beaker-rspec` testing.
2. `.fixtures.yml` is deprecated. If it the file exists, it is the sole source of mods for `rspec-puppet` and `metadata.json` is ignored.
3. If the dep data in `metadata.json` is insufficient, for example as a git ref needs to be used, a `Puppetfile` may be placed in  he root of the module.  An equivalent to `Gemfile`'s `gemspec` in the `Puppetfile` would be useful but initially it could somply be that the presence of `./Puppetfile` would cause it to become the sole mod inventory.
4. An equivalent `Puppetfile` for `beaker-rspec` could be `./spec/support/acceptance/Puppetfile`.

Reply:
In Foreman we actually use `.fixtures.yml` also for our acceptance tests. This means we actually get git versions there too, which has been useful. For example, we have a puppet-certs module and this is used in another. We will depend on the next version, but to avoid doing 3 major releases in one cycle it allows us to move forward faster.

I feel this is a bit of an obscure place to have the discussion. We're really talking about fundamental module design/layout and that deserves wider attention. Would the mailing list be a better place?